src/core-sdk-tasks/DownloadFile.cs

```diff
@@ -39,31 +54,75 @@ public override bool Execute()
                 {
-                    var getTask = httpClient.GetStreamAsync(Uri);
+                    downloadStatus = await DownloadWithRetriesAsync(PrivateUri, DestinationPath);
+                }
+            }
+
+            if (downloadStatus != true)
+            {
+                throw new FileNotFoundException($"Error while attempting download file from the provided URLs.");
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Attempt to download file from `source` with retries when response error is different of FileNotFound and Success.
+        /// </summary>
+        /// <param name="source">URL to the file to be downloaded.</param>
+        /// <param name="target">Local path where to put the downloaded file.</param>
+        /// <returns>true: Download Succeeded. false: Download failed with 404. null: Download failed but is retriable.</returns>
+        private async Task<bool?> DownloadWithRetriesAsync(string source, string target)
+        {
+            Random rng = new Random();
 
+            Log.LogMessage(MessageImportance.High, $"Attempting download '{source}' to '{target}'");
+
+            using (var httpClient = new HttpClient())
+            {
+                for (int retryNumber = 0; retryNumber < MaxRetries; retryNumber++)
+                {
                     try
                     {
-                        using (var outStream = File.Create(DestinationPath))
+                        var httpResponse = await httpClient.GetAsync(source);
+
+                        // The Azure Storage REST API returns '400 - Bad Request' in some cases
+                        // where the resource is not found on the storage.
+                        // https://docs.microsoft.com/en-us/rest/api/storageservices/common-rest-api-error-codes
+                        if (httpResponse.StatusCode == HttpStatusCode.NotFound ||
+                            httpResponse.ReasonPhrase.IndexOf("The requested URI does not represent any resource on the server.", StringComparison.OrdinalIgnoreCase) >= 0)
+                        {
+                            Log.LogMessage($"File not found error while attempting download using the following URL(s): '{source}'.");
+                            return false;
+                        }
+
+                        using (var outStream = File.Create(target))
                         {
-                            getTask.Result.CopyTo(outStream);
+                            await httpResponse.Content.CopyToAsync(outStream);
                         }
+
+                        return true;
                     }
-                    catch (Exception)
+                    catch (Exception e)
                     {
-                        File.Delete(DestinationPath);
-                        throw;
+                        Log.LogMessage($"Problems downloading file from '{source}'. {e.Message} {e.StackTrace}");
```

Comment:
This message also looks like it'll still only show up in the binlog.

Reply:
Changed it.